vms/proposervm/proposer/windower.go

```diff
@@ -148,3 +168,125 @@ func (w *windower) Delay(ctx context.Context, chainHeight, pChainHeight uint64,
 	}
 	return delay, nil
 }
+
+func (w *windower) ExpectedProposer(
+	ctx context.Context,
+	blockHeight,
+	pChainHeight,
+	slot uint64,
+) (ids.NodeID, error) {
+	validators, err := w.sortedValidators(ctx, pChainHeight)
+	if err != nil {
+		return ids.EmptyNodeID, err
+	}
+
+	var (
+		seed             = postDurangoSeed(w.chainSource, blockHeight, slot)
+		source           = prng.NewMT19937_64()
+		validatorWeights = validatorsToWeight(validators)
+	)
+
+	sampler := sampler.NewDeterministicWeightedWithoutReplacement(source)
+	if err := sampler.Initialize(validatorWeights); err != nil {
+		return ids.EmptyNodeID, err
+	}
+
+	source.Seed(seed)
+	indices, err := sampler.Sample(1)
+	if err != nil {
+		return ids.EmptyNodeID, fmt.Errorf("%w, %w", err, ErrNoProposersAvailable)
+	}
+	return validators[indices[0]].id, nil
+}
+
+func (w *windower) MinDelayForProposer(
```

Comment:
nit: I think if this accepted `maxSlot` as an argument, we could call this from `ExpectedProposer`

Reply:
Hmmm I'm not sure how

Reply:
More or less (or just make `ExpectedProposer` a very thin wrapper around it).

Just optional as a way to reduce related code usage.

Reply:
I'm still not sure exactly what you think the code should look like (`ExpectedProposer` is returning the `ids.NodeID` which isn't provided by this function)... But I restructured the helpers to avoid as much duplication as I could.

Reply:
This is what I had in mind.